azure-self-hosted-runners/post-deployment-script.ps1

```diff
@@ -0,0 +1,121 @@
+Write-Output "Finished adding Microsoft Defender Exclusions."
+
+# ======================
+# GIT FOR WINDOWS
+# ======================
+
+Write-Output "Downloading Git for Windows..."
+$GitForWindowsOutputFile = "./git-for-windows-installer.exe"
+$ProgressPreference = 'SilentlyContinue'
+Invoke-WebRequest -UseBasicParsing -Uri "https://github.com/git-for-windows/git/releases/download/v${GitForWindowsTag}/Git-${GitForWindowsVersion}-64-bit.exe" -OutFile $GitForWindowsOutputFile
+$ProgressPreference = 'Continue'
+
+if((Get-FileHash -Path $GitForWindowsOutputFile -Algorithm SHA256).Hash.ToUpper() -ne $GitForWindowsHash.ToUpper()){ throw 'Computed checksum did not match' }
+
+Write-Output "Installing Git for Windows..."
+@"
+[Setup]
+Lang=default
+Dir=C:\Program Files\Git
+Group=Git
+NoIcons=0
+SetupType=default
+Components=gitlfs,windowsterminal
+Tasks=
+EditorOption=VIM
+CustomEditorPath=
+DefaultBranchOption= 
+PathOption=CmdTools
+SSHOption=OpenSSH
+TortoiseOption=false
+CURLOption=WinSSL
+CRLFOption=CRLFAlways
+BashTerminalOption=ConHost
+GitPullBehaviorOption=FFOnly
+UseCredentialManager=Core
+PerformanceTweaksFSCache=Enabled
+EnableSymlinks=Disabled
+EnablePseudoConsoleSupport=Disabled
+EnableFSMonitor=Disabled
+"@ | Out-File -FilePath "./git-installer-config.inf"
+
+Start-Process -Wait $GitForWindowsOutputFile '/VERYSILENT /NORESTART /NOCANCEL /SP- /CLOSEAPPLICATIONS /RESTARTAPPLICATIONS /LOADINF="./git-installer-config.inf"'
+
+Write-Output "Finished installing Git for Windows."
+
+# ======================
+# GITHUB ACTIONS RUNNER
+# ======================
+
+Write-Output "Downloading GitHub Actions runner..."
+
+mkdir $GitHubActionsRunnerPath | Out-Null
+$ProgressPreference = 'SilentlyContinue'
+Invoke-WebRequest -UseBasicParsing -Uri https://github.com/actions/runner/releases/download/v${GitHubActionsRunnerVersion}/actions-runner-win-${GithubActionsRunnerArch}-${GitHubActionsRunnerVersion}.zip -OutFile ${GitHubActionsRunnerPath}\actions-runner-win-${GithubActionsRunnerArch}-${GitHubActionsRunnerVersion}.zip
+$ProgressPreference = 'Continue'
+if((Get-FileHash -Path ${GitHubActionsRunnerPath}\actions-runner-win-${GithubActionsRunnerArch}-${GitHubActionsRunnerVersion}.zip -Algorithm SHA256).Hash.ToUpper() -ne $GithubActionsRunnerHash.ToUpper()){ throw 'Computed checksum did not match' }
+
+Write-Output "Installing GitHub Actions runner ${GitHubActionsRunnerVersion} as a Windows service with labels ${GithubActionsRunnerLabels}..."
+
+Add-Type -AssemblyName System.IO.Compression.FileSystem ; [System.IO.Compression.ZipFile]::ExtractToDirectory("${GitHubActionsRunnerPath}\actions-runner-win-${GithubActionsRunnerArch}-${GitHubActionsRunnerVersion}.zip", $GitHubActionsRunnerPath)
+cmd.exe /c "${GitHubActionsRunnerPath}\config.cmd" --unattended --ephemeral --name ${GithubActionsRunnerName} --runasservice --labels ${GithubActionsRunnerLabels} --url ${GithubActionsRunnerRegistrationUrl} --token ${GitHubActionsRunnerToken}
```

Comment:
I added `--ephemeral` so that the VM will only run a single GitHub Actions job. We'll need to add some logic to shutdown and remove the VM after use. I'm thinking about something like:
- Setting the [brand-new](https://docs.github.com/en/actions/hosting-your-own-runners/running-scripts-before-or-after-a-job) `ACTIONS_RUNNER_HOOK_JOB_COMPLETED` env variable on the VM that runs a script after the job has finished. This script could:
- From that script, call the [`Azure Instance Metadata Service`](https://learn.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows) on `169.254.169.254` to figure out the name of the currently running VM. Something like this should work:
 
```PowerShell
$Proxy=New-object System.Net.WebProxy
$WebSession=new-object Microsoft.PowerShell.Commands.WebRequestSession
$WebSession.Proxy=$Proxy
$Data=Invoke-RestMethod -Headers @{"Metadata"="true"} -Method GET -Uri "http://169.254.169.254/metadata/instance?api-version=2021-02-01" -WebSession $WebSession | ConvertTo-Json -Depth 64
$ParsedData = $Data| ConvertFrom-Json
echo $ParsedData.compute.name
# Returns e.g. actions-runner-20221220130445797318063
echo $ParsedData.compute.resourceGroupName
# Returns e.g. github-actions-runners
echo $ParsedData.compute.resourceId
# Returns e.g. /subscriptions/YOUR_SUB_ID_HERE/resourceGroups/github-actions-runners/providers/Microsoft.Compute/virtualMachines/actions-runner-20221220130445797318063
```

- The script should then trigger some sort of webhook call to request removal of the VM by the GfW bot, given the name and resource group name (or resourceId) we just got from the metadata endpoint. WDYT?

Reply:
Sounds good!

FWIW I thought that the GitForWindowsHelper GitHub App would receive a webhook event after the workflow run completed that would contain enough information so that we could remove the entire resource group (which I hoped would automatically stop and delete the VM). Here is an example for a recent delivery of such a `check_run.completed` webhook event:

<details><summary>Headers</summary>

```yaml
Request method: POST
Accept: */*
content-type: application/json
User-Agent: GitHub-Hookshot/2179efe
X-GitHub-Delivery: 290d2ca0-8141-11ed-98cb-b0a992770f94
X-GitHub-Event: check_run
X-GitHub-Hook-ID: 392499201
X-GitHub-Hook-Installation-Target-ID: 246505
X-GitHub-Hook-Installation-Target-Type: integration
X-Hub-Signature: sha1=85a088f44616d331bf60f6c119727d5e81b95aa6
X-Hub-Signature-256: sha256=71e60249b346e085b93df13a43a79237ac3db4166d7486eaeea83b913af59151
```

</details>

<details><summary>Payload</summary>

```yaml
{
  "action": "completed",
  "check_run": {
    "id": 10233939147,
    "name": "win+VS build (arm64)",
    "node_id": "CR_kwDOAWJAkM8AAAACYf2Eyw",
    "head_sha": "6a20512f99c2a5012d0972d1f926a835730c274d",
    "external_id": "e1b98089-497b-54b4-42e0-e2656d08cc2a",
    "url": "https://api.github.com/repos/git-for-windows/git/check-runs/10233939147",
    "html_url": "https://github.com/git-for-windows/git/actions/runs/3750386564/jobs/6370043175",
    "details_url": "https://github.com/git-for-windows/git/actions/runs/3750386564/jobs/6370043175",
    "status": "completed",
    "conclusion": "success",
    "started_at": "2022-12-21T15:01:09Z",
    "completed_at": "2022-12-21T15:07:16Z",
    "output": {
      "title": null,
      "summary": null,
      "text": null,
      "annotations_count": 0,
      "annotations_url": "https://api.github.com/repos/git-for-windows/git/check-runs/10233939147/annotations"
    },
    "check_suite": {
      "id": 9993156801,
      "node_id": "CS_kwDOAWJAkM8AAAACU6N4wQ",
      "head_branch": "shears/main",
      "head_sha": "6a20512f99c2a5012d0972d1f926a835730c274d",
      "status": "queued",
      "conclusion": null,
      "url": "https://api.github.com/repos/git-for-windows/git/check-suites/9993156801",
      "before": "3dd6ea96c600566c590fb3eedfde352fca5b7067",
      "after": "6a20512f99c2a5012d0972d1f926a835730c274d",
      "pull_requests": [

      ],
      "app": {
        "id": 15368,
        "slug": "github-actions",
        "node_id": "MDM6QXBwMTUzNjg=",
        "owner": {
          "login": "github",
          "id": 9919,
          "node_id": "MDEyOk9yZ2FuaXphdGlvbjk5MTk=",
          "avatar_url": "https://avatars.githubusercontent.com/u/9919?v=4",
          "gravatar_id": "",
          "url": "https://api.github.com/users/github",
          "html_url": "https://github.com/github",
          "followers_url": "https://api.github.com/users/github/followers",
          "following_url": "https://api.github.com/users/github/following{/other_user}",
          "gists_url": "https://api.github.com/users/github/gists{/gist_id}",
          "starred_url": "https://api.github.com/users/github/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/github/subscriptions",
          "organizations_url": "https://api.github.com/users/github/orgs",
          "repos_url": "https://api.github.com/users/github/repos",
          "events_url": "https://api.github.com/users/github/events{/privacy}",
          "received_events_url": "https://api.github.com/users/github/received_events",
          "type": "Organization",
          "site_admin": false
        },
        "name": "GitHub Actions",
        "description": "Automate your workflow from idea to production",
        "external_url": "https://help.github.com/en/actions",
        "html_url": "https://github.com/apps/github-actions",
        "created_at": "2018-07-30T09:30:17Z",
        "updated_at": "2019-12-10T19:04:12Z",
        "permissions": {
          "actions": "write",
          "administration": "read",
          "checks": "write",
          "contents": "write",
          "deployments": "write",
          "discussions": "write",
          "issues": "write",
          "merge_queues": "write",
          "metadata": "read",
          "packages": "write",
          "pages": "write",
          "pull_requests": "write",
          "repository_hooks": "write",
          "repository_projects": "write",
          "security_events": "write",
          "statuses": "write",
          "vulnerability_alerts": "read"
        },
        "events": [
          "branch_protection_rule",
          "check_run",
          "check_suite",
          "create",
          "delete",
          "deployment",
          "deployment_status",
          "discussion",
          "discussion_comment",
          "fork",
          "gollum",
          "issues",
          "issue_comment",
          "label",
          "merge_group",
          "milestone",
          "page_build",
          "project",
          "project_card",
          "project_column",
          "public",
          "pull_request",
          "pull_request_review",
          "pull_request_review_comment",
          "push",
          "registry_package",
          "release",
          "repository",
          "repository_dispatch",
          "status",
          "watch",
          "workflow_dispatch",
          "workflow_run"
        ]
      },
      "created_at": "2022-12-21T15:00:34Z",
      "updated_at": "2022-12-21T15:07:17Z"
    },
    "app": {
      "id": 15368,
      "slug": "github-actions",
      "node_id": "MDM6QXBwMTUzNjg=",
      "owner": {
        "login": "github",
        "id": 9919,
        "node_id": "MDEyOk9yZ2FuaXphdGlvbjk5MTk=",
        "avatar_url": "https://avatars.githubusercontent.com/u/9919?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/github",
        "html_url": "https://github.com/github",
        "followers_url": "https://api.github.com/users/github/followers",
        "following_url": "https://api.github.com/users/github/following{/other_user}",
        "gists_url": "https://api.github.com/users/github/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/github/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/github/subscriptions",
        "organizations_url": "https://api.github.com/users/github/orgs",
        "repos_url": "https://api.github.com/users/github/repos",
        "events_url": "https://api.github.com/users/github/events{/privacy}",
        "received_events_url": "https://api.github.com/users/github/received_events",
        "type": "Organization",
        "site_admin": false
      },
      "name": "GitHub Actions",
      "description": "Automate your workflow from idea to production",
      "external_url": "https://help.github.com/en/actions",
      "html_url": "https://github.com/apps/github-actions",
      "created_at": "2018-07-30T09:30:17Z",
      "updated_at": "2019-12-10T19:04:12Z",
      "permissions": {
        "actions": "write",
        "administration": "read",
        "checks": "write",
        "contents": "write",
        "deployments": "write",
        "discussions": "write",
        "issues": "write",
        "merge_queues": "write",
        "metadata": "read",
        "packages": "write",
        "pages": "write",
        "pull_requests": "write",
        "repository_hooks": "write",
        "repository_projects": "write",
        "security_events": "write",
        "statuses": "write",
        "vulnerability_alerts": "read"
      },
      "events": [
        "branch_protection_rule",
        "check_run",
        "check_suite",
        "create",
        "delete",
        "deployment",
        "deployment_status",
        "discussion",
        "discussion_comment",
        "fork",
        "gollum",
        "issues",
        "issue_comment",
        "label",
        "merge_group",
        "milestone",
        "page_build",
        "project",
        "project_card",
        "project_column",
        "public",
        "pull_request",
        "pull_request_review",
        "pull_request_review_comment",
        "push",
        "registry_package",
        "release",
        "repository",
        "repository_dispatch",
        "status",
        "watch",
        "workflow_dispatch",
        "workflow_run"
      ]
    },
    "pull_requests": [

    ]
  },
  "repository": {
    "id": 23216272,
    "node_id": "MDEwOlJlcG9zaXRvcnkyMzIxNjI3Mg==",
    "name": "git",
    "full_name": "git-for-windows/git",
    "private": false,
    "owner": {
      "login": "git-for-windows",
      "id": 4571183,
      "node_id": "MDEyOk9yZ2FuaXphdGlvbjQ1NzExODM=",
      "avatar_url": "https://avatars.githubusercontent.com/u/4571183?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/git-for-windows",
      "html_url": "https://github.com/git-for-windows",
      "followers_url": "https://api.github.com/users/git-for-windows/followers",
      "following_url": "https://api.github.com/users/git-for-windows/following{/other_user}",
      "gists_url": "https://api.github.com/users/git-for-windows/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/git-for-windows/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/git-for-windows/subscriptions",
      "organizations_url": "https://api.github.com/users/git-for-windows/orgs",
      "repos_url": "https://api.github.com/users/git-for-windows/repos",
      "events_url": "https://api.github.com/users/git-for-windows/events{/privacy}",
      "received_events_url": "https://api.github.com/users/git-for-windows/received_events",
      "type": "Organization",
      "site_admin": false
    },
    "html_url": "https://github.com/git-for-windows/git",
    "description": "A fork of Git containing Windows-specific patches.",
    "fork": true,
    "url": "https://api.github.com/repos/git-for-windows/git",
    "forks_url": "https://api.github.com/repos/git-for-windows/git/forks",
    "keys_url": "https://api.github.com/repos/git-for-windows/git/keys{/key_id}",
    "collaborators_url": "https://api.github.com/repos/git-for-windows/git/collaborators{/collaborator}",
    "teams_url": "https://api.github.com/repos/git-for-windows/git/teams",
    "hooks_url": "https://api.github.com/repos/git-for-windows/git/hooks",
    "issue_events_url": "https://api.github.com/repos/git-for-windows/git/issues/events{/number}",
    "events_url": "https://api.github.com/repos/git-for-windows/git/events",
    "assignees_url": "https://api.github.com/repos/git-for-windows/git/assignees{/user}",
    "branches_url": "https://api.github.com/repos/git-for-windows/git/branches{/branch}",
    "tags_url": "https://api.github.com/repos/git-for-windows/git/tags",
    "blobs_url": "https://api.github.com/repos/git-for-windows/git/git/blobs{/sha}",
    "git_tags_url": "https://api.github.com/repos/git-for-windows/git/git/tags{/sha}",
    "git_refs_url": "https://api.github.com/repos/git-for-windows/git/git/refs{/sha}",
    "trees_url": "https://api.github.com/repos/git-for-windows/git/git/trees{/sha}",
    "statuses_url": "https://api.github.com/repos/git-for-windows/git/statuses/{sha}",
    "languages_url": "https://api.github.com/repos/git-for-windows/git/languages",
    "stargazers_url": "https://api.github.com/repos/git-for-windows/git/stargazers",
    "contributors_url": "https://api.github.com/repos/git-for-windows/git/contributors",
    "subscribers_url": "https://api.github.com/repos/git-for-windows/git/subscribers",
    "subscription_url": "https://api.github.com/repos/git-for-windows/git/subscription",
    "commits_url": "https://api.github.com/repos/git-for-windows/git/commits{/sha}",
    "git_commits_url": "https://api.github.com/repos/git-for-windows/git/git/commits{/sha}",
    "comments_url": "https://api.github.com/repos/git-for-windows/git/comments{/number}",
    "issue_comment_url": "https://api.github.com/repos/git-for-windows/git/issues/comments{/number}",
    "contents_url": "https://api.github.com/repos/git-for-windows/git/contents/{+path}",
    "compare_url": "https://api.github.com/repos/git-for-windows/git/compare/{base}...{head}",
    "merges_url": "https://api.github.com/repos/git-for-windows/git/merges",
    "archive_url": "https://api.github.com/repos/git-for-windows/git/{archive_format}{/ref}",
    "downloads_url": "https://api.github.com/repos/git-for-windows/git/downloads",
    "issues_url": "https://api.github.com/repos/git-for-windows/git/issues{/number}",
    "pulls_url": "https://api.github.com/repos/git-for-windows/git/pulls{/number}",
    "milestones_url": "https://api.github.com/repos/git-for-windows/git/milestones{/number}",
    "notifications_url": "https://api.github.com/repos/git-for-windows/git/notifications{?since,all,participating}",
    "labels_url": "https://api.github.com/repos/git-for-windows/git/labels{/name}",
    "releases_url": "https://api.github.com/repos/git-for-windows/git/releases{/id}",
    "deployments_url": "https://api.github.com/repos/git-for-windows/git/deployments",
    "created_at": "2014-08-22T07:07:36Z",
    "updated_at": "2022-12-21T08:24:32Z",
    "pushed_at": "2022-12-21T15:00:55Z",
    "git_url": "git://github.com/git-for-windows/git.git",
    "ssh_url": "git@github.com:git-for-windows/git.git",
    "clone_url": "https://github.com/git-for-windows/git.git",
    "svn_url": "https://github.com/git-for-windows/git",
    "homepage": "http://git-scm.com/",
    "size": 267177,
    "stargazers_count": 7204,
    "watchers_count": 7204,
    "language": "C",
    "has_issues": true,
    "has_projects": true,
    "has_downloads": true,
    "has_wiki": true,
    "has_pages": false,
    "has_discussions": true,
    "forks_count": 2306,
    "mirror_url": null,
    "archived": false,
    "disabled": false,
    "open_issues_count": 151,
    "license": {
      "key": "other",
      "name": "Other",
      "spdx_id": "NOASSERTION",
      "url": null,
      "node_id": "MDc6TGljZW5zZTA="
    },
    "allow_forking": true,
    "is_template": false,
    "web_commit_signoff_required": false,
    "topics": [
      "hacktoberfest"
    ],
    "visibility": "public",
    "forks": 2306,
    "open_issues": 151,
    "watchers": 7204,
    "default_branch": "main"
  },
  "organization": {
    "login": "git-for-windows",
    "id": 4571183,
    "node_id": "MDEyOk9yZ2FuaXphdGlvbjQ1NzExODM=",
    "url": "https://api.github.com/orgs/git-for-windows",
    "repos_url": "https://api.github.com/orgs/git-for-windows/repos",
    "events_url": "https://api.github.com/orgs/git-for-windows/events",
    "hooks_url": "https://api.github.com/orgs/git-for-windows/hooks",
    "issues_url": "https://api.github.com/orgs/git-for-windows/issues",
    "members_url": "https://api.github.com/orgs/git-for-windows/members{/member}",
    "public_members_url": "https://api.github.com/orgs/git-for-windows/public_members{/member}",
    "avatar_url": "https://avatars.githubusercontent.com/u/4571183?v=4",
    "description": ""
  },
  "sender": {
    "login": "git-for-windows-ci",
    "id": 24522801,
    "node_id": "MDQ6VXNlcjI0NTIyODAx",
    "avatar_url": "https://avatars.githubusercontent.com/u/24522801?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/git-for-windows-ci",
    "html_url": "https://github.com/git-for-windows-ci",
    "followers_url": "https://api.github.com/users/git-for-windows-ci/followers",
    "following_url": "https://api.github.com/users/git-for-windows-ci/following{/other_user}",
    "gists_url": "https://api.github.com/users/git-for-windows-ci/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/git-for-windows-ci/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/git-for-windows-ci/subscriptions",
    "organizations_url": "https://api.github.com/users/git-for-windows-ci/orgs",
    "repos_url": "https://api.github.com/users/git-for-windows-ci/repos",
    "events_url": "https://api.github.com/users/git-for-windows-ci/events{/privacy}",
    "received_events_url": "https://api.github.com/users/git-for-windows-ci/received_events",
    "type": "User",
    "site_admin": false
  },
  "installation": {
    "id": 31544731,
    "node_id": "MDIzOkludGVncmF0aW9uSW5zdGFsbGF0aW9uMzE1NDQ3MzE="
  }
}
```

</details>

I guess there is not quite enough information to tell whether this was run on a self-hosted runner, basically because we only get a `check_run` which is not a `job` that would have been run on a particular runner. But I suspect the relevant information might only be a call to e.g. https://docs.github.com/en/rest/actions/workflow-runs#get-a-workflow-run away, maybe also https://docs.github.com/en/rest/actions/workflow-jobs#get-a-job-for-a-workflow-run would be needed.